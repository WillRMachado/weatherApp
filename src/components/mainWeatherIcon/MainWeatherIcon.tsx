import React from 'react';
import {StyleSheet, View} from 'react-native';
import {useTheme} from '@react-navigation/native';
import {themeColorsTypes} from '../../styles/themes/themesType';
import Feather from 'react-native-vector-icons/Feather';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import {useSelector} from 'react-redux';

const getIconComponent = (
  main: string,
  size: number,
  color: string,
  isDark?: boolean,
) => {
  switch (main) {
    case 'Thunderstorm':
      return <Feather name={'cloud-lightning'} size={size} color={color} />;
    case 'Drizzle':
      return <Feather name={'cloud-drizzle'} size={size} color={color} />;
    case 'Rain':
      return <Feather name={'cloud-rain'} size={size} color={color} />;
    case 'Snow':
      return <Feather name={'cloud-snow'} size={size} color={color} />;
    case 'Tornado':
      return (
        <MaterialCommunityIcons
          name={'weather-tornado'}
          size={size}
          color={color}
        />
      );
    case 'Clear':
      if (isDark) {
        return <Feather name={'moon'} size={size} color={color} />;
      } else {
        return <Feather name={'sun'} size={size} color={color} />;
      }
    case 'Clouds':
      return <Feather name={'cloud'} size={size} color={color} />;
    case 'Mist':
    case 'Smoke':
    case 'Haze':
    case 'Dust':
    case 'Fog':
    case 'Sand':
    case 'Ash':
    case 'Ash':
    case 'Squall':
      return (
        <MaterialCommunityIcons
          name={'weather-fog'}
          size={size}
          color={color}
        />
      );
    default:
      return <Feather name={'frown'} size={size} color={color} />;
  }
};

export default function MainWeatherIcon(props: {
  main: string;
  iconSize?: number;
  style?: object;
}) {
  const {main, style, iconSize} = props;
  const {colors}: themeColorsTypes = useTheme();
  const isDark = useSelector(
    (state: any) => state.store.weather.current.isDark,
  );

  return (
    <View style={[styles.defaultStyle, {...style}]}>
      {getIconComponent(main, iconSize || 90, colors.secondary, isDark)}
    </View>
  );
}

const styles = StyleSheet.create({
  defaultStyle: {
    width: 100,
    height: 100,
    justifyContent: 'center',
    alignItems: 'center',
  },
});
