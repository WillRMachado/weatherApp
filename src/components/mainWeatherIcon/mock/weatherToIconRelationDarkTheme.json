[
  {
    "weatherName": "Thunderstorm",
    "iconsFamily": "Feather",
    "iconName": "cloud-lightning"
  },
  {
    "weatherName": "Drizzle",
    "iconsFamily": "Feather",
    "iconName": "cloud-drizzle"
  },
  {
    "weatherName": "Rain",
    "iconsFamily": "Feather",
    "iconName": "cloud-rain"
  },
  {
    "weatherName": "Snow",
    "iconsFamily": "Feather",
    "iconName": "cloud-snow"
  },
  {
    "weatherName": "Tornado",
    "iconsFamily": "MaterialCommunityIcons",
    "iconName": "weather-tornado"
  },
  {
    "weatherName": "Clear",
    "iconsFamily": "Feather",
    "iconName": "moon"
  },
  {
    "weatherName": "Clouds",
    "iconsFamily": "Feather",
    "iconName": "cloud"
  },
  {
    "weatherName": "Mist",
    "iconsFamily": "MaterialCommunityIcons",
    "iconName": "weather-fog"
  },
  {
    "weatherName": "Smoke",
    "iconsFamily": "MaterialCommunityIcons",
    "iconName": "weather-fog"
  },
  {
    "weatherName": "Haze",
    "iconsFamily": "MaterialCommunityIcons",
    "iconName": "weather-fog"
  },
  {
    "weatherName": "Dust",
    "iconsFamily": "MaterialCommunityIcons",
    "iconName": "weather-fog"
  },
  {
    "weatherName": "Fog",
    "iconsFamily": "MaterialCommunityIcons",
    "iconName": "weather-fog"
  },
  {
    "weatherName": "Sand",
    "iconsFamily": "MaterialCommunityIcons",
    "iconName": "weather-fog"
  },
  {
    "weatherName": "Ash",
    "iconsFamily": "MaterialCommunityIcons",
    "iconName": "weather-fog"
  },
  {
    "weatherName": "Squall",
    "iconsFamily": "MaterialCommunityIcons",
    "iconName": "weather-fog"
  }
]
