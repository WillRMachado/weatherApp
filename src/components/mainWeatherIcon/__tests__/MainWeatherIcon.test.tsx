import React from 'react';
import screenRender from '../../../../jestHelpers/screenRender';
import screenSnapshotCreator from '../../../../jestHelpers/screenSnapshotCreator';
import MainWeatherIcon from '../MainWeatherIcon';
import Feather from 'react-native-vector-icons/Feather';
import MaterialCommunityIcons from 'react-native-vector-icons/MaterialCommunityIcons';
import weatherToIconRelationLightTheme from '../mock/weatherToIconRelationLightTheme.json';
import weatherToIconRelationDarkTheme from '../mock/weatherToIconRelationDarkTheme.json';
import * as reactRedux from 'react-redux';

describe('MainWeatherIcon Component', () => {
  const mockWeather = 'Clear';
  const useSelectorMock = jest.spyOn(reactRedux, 'useSelector');

  const MainWeatherIconWithProps = () => <MainWeatherIcon main={mockWeather} />;

  beforeEach(() => {
    useSelectorMock.mockClear();
  });

  it('renders correctly', async () => {
    const component = await screenRender(MainWeatherIconWithProps);
    expect(component).toBeTruthy();
  });

  it('matches snapshot', async () => {
    const component = await screenSnapshotCreator(MainWeatherIconWithProps);
    expect(component).toMatchSnapshot();
  });

  it('gets weather icon properly on light theme', async () => {
    for (
      let index = 0;
      index < weatherToIconRelationLightTheme.length;
      index++
    ) {
      const element = weatherToIconRelationLightTheme[index];

      const MainWeatherIconWithCurrentIcon = () => (
        <MainWeatherIcon main={element.weatherName} />
      );
      const component = await screenRender(MainWeatherIconWithCurrentIcon);

      let renderedIcon;
      if (element.iconsFamily === 'Feather') {
        renderedIcon = component.root.findAllByType(Feather)[0];
      } else if (element.iconsFamily === 'MaterialCommunityIcons') {
        renderedIcon = component.root.findAllByType(MaterialCommunityIcons)[0];
      } else {
        throw new Error('unidentified icon family');
      }

      const renderedIconName = renderedIcon.props.name;

      expect(renderedIconName).toEqual(element.iconName);
    }
  });

  it('gets weather icon properly on dark theme', async () => {
    useSelectorMock.mockReturnValue(true);

    for (
      let index = 0;
      index < weatherToIconRelationDarkTheme.length;
      index++
    ) {
      const element = weatherToIconRelationDarkTheme[index];

      const MainWeatherIconWithCurrentIcon = () => (
        <MainWeatherIcon main={element.weatherName} />
      );
      const component = await screenRender(MainWeatherIconWithCurrentIcon);

      let renderedIcon;
      if (element.iconsFamily === 'Feather') {
        renderedIcon = component.root.findAllByType(Feather)[0];
      } else if (element.iconsFamily === 'MaterialCommunityIcons') {
        renderedIcon = component.root.findAllByType(MaterialCommunityIcons)[0];
      } else {
        throw new Error('unidentified icon family');
      }

      const renderedIconName = renderedIcon.props.name;

      expect(renderedIconName).toEqual(element.iconName);
    }
  });
});
