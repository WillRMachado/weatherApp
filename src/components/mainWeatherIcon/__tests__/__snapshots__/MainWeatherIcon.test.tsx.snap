// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MainWeatherIcon Component matches snapshot 1`] = `null`;
